pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java

```diff
@@ -357,16 +357,24 @@ private void updateAllBrokerData() {
                         // Replace previous local broker data.
                         brokerDataMap.get(broker).setLocalData(localData);
                     } else {
-                        // Initialize BrokerData object for previously unseen
-                        // brokers.
+                        // Initialize BrokerData object for previously unseen brokers.
                         brokerDataMap.put(broker, new BrokerData(localData));
                     }
                 } catch (Exception e) {
                     log.warn("Error reading broker data from cache for broker - [{}], [{}]", broker, e.getMessage());
                 }
             }
+            // Remove obsolete brokers.
+            for (final String broker : brokerDataMap.keySet()) {
```

Comment:
Actually, updating `brokerDataMap` will still not fix issue 100% because `updateAllBrokerData` being called async when broker receives the watch. So, before it gets updated, lookup may select broker and it fails to get broker-url while creating lookup result.
So, while selecting broker at [selectBrokerForAssignment](https://github.com/bobbeyreese/pulsar/blob/6e491f3bad333d4c61ea1074ec3f6b9f6adddb45/pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/ModularLoadManagerImpl.java#L504) it should read directly from the cache (as per your previous change) so, in that way cache is consistent in terms of selecting-broker and creating-lookupResult.